### PR TITLE
GroupDataProvider: PersistentStorageDelegate Initialization

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
The storage delegate should be injected using the Init() method, instead of the constructor
* Relates to [#16028](https://github.com/project-chip/connectedhomeip/issues/16028)

#### Change overview
* The storage delegate is now an argument of the GroupDataProvider::Init()
* Group data provider implementation updated accordingly
* Calls to the GroupDataProvider constructors, and Init() updated accordingly

#### Testing
Group, and Group Data Provider unit tests and YAML tests ran successfully.

### CODE REVIEW

File: src/credentials/GroupDataProvider.h [hunk 1]
```diff
@@ -227,8 +228,8 @@ class GroupDataProvider
      *  @retval #CHIP_ERROR_INCORRECT_STATE if called when already initialized.
      *  @retval #CHIP_NO_ERROR on success
      */
-    virtual CHIP_ERROR Init() = 0;
-    virtual void Finish()     = 0;
+    virtual CHIP_ERROR Init(PersistentStorageDelegate * storage) = 0;
```
Comment (tcarmelveilleux) [inline src/credentials/GroupDataProvider.h:231]: This init should not rely on `PersistentStorageDelegate`.

Suggest that we keep init no-params here, and have `GroupDataProviderImpl::SetPersistentStorageDelegate(PersistentStorageDelegate *)`.

Then `GroupDataProviderImpl::Init()`  could just fail with CHIP_ERROR_INCORRECT_STATE if `SetPersistentStorageDelegate` not called.

File: src/credentials/GroupDataProviderImpl.h [hunk 2]
```diff
@@ -215,8 +212,8 @@ class GroupDataProviderImpl : public GroupDataProvider
     };
     CHIP_ERROR RemoveEndpoints(FabricIndex fabric_index, GroupId group_id);
 
-    chip::PersistentStorageDelegate & mStorage;
-    bool mInitialized = false;
+    chip::PersistentStorageDelegate * mStorage = nullptr;
+    bool mInitialized                          = false;
```
Comment (tcarmelveilleux) [inline src/credentials/GroupDataProviderImpl.h:216]: Can be replaced by the fact that `bool IsInitialized() { return (mStorage == nullptr); }`

File: src/app/server/Server.cpp [hunk 3]
```diff
@@ -138,7 +138,7 @@ CHIP_ERROR Server::Init(AppDelegate * delegate, uint16_t secureServicePort, uint
     app::DnssdServer::Instance().SetCommissioningModeProvider(&mCommissioningWindowManager);
 
     // Group data provider must be initialized after mDeviceStorage
-    err = mGroupsProvider.Init();
+    err = mGroupsProvider.Init(&mDeviceStorage);
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:141]: I don't see how this fixes https://github.com/project-chip/connectedhomeip/issues/16028, which is not even about the groups data provider as far as I can tell.  What am I missing?

Comment (tcarmelveilleux) [reply src/app/server/Server.cpp:141]: It doesn't fix 16028. I think it's a mistake the fixes line in the description. The clean-up of Server dependencies to storage is on the way to single storage injection points for Server and controller, so it relates to 16028 but doesn't fix it

Comment (rcasallas-silabs) [reply src/app/server/Server.cpp]: You're both correct. I have replace the word "Fix" with "Relates" in the description.

File: src/credentials/GroupDataProviderImpl.h (hunk 2 above)
Comment (rcasallas-silabs) [reply src/credentials/GroupDataProviderImpl.h]: Fixed as suggested.

File: src/credentials/GroupDataProvider.h (hunk 1 above)
Comment (rcasallas-silabs) [reply src/credentials/GroupDataProvider.h]: I have removed the parameter from the `Init` method, and replaced with a new `SetStorageDelegate`.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/credentials/GroupDataProvider.h [hunk 4]
```diff
@@ -221,8 +221,9 @@ class GroupDataProvider
     uint16_t GetMaxGroupKeysPerFabric() { return mMaxGroupKeysPerFabric; }
 
     /**
-     *  Initialize the GroupDataProvider, including any persistent data store
-     *  initialization. Must be called once before any other API succeeds.
+     *  Initialize the GroupDataProvider, including possilby any persistent
```
Comment (bzbarsky-apple) [inline src/credentials/GroupDataProvider.h:224]:
```suggestion
     *  Initialize the GroupDataProvider, including possibly any persistent
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by emargolis — APPROVED** [review]

**Pull request merged**